internal/exp/metrics/streams/streams.go

```diff
@@ -0,0 +1,88 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package streams // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/streams"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+)
+
+// Sequence of streams that can be iterated upon
+type Seq[T any] func(yield func(identity.Stream, T) bool) bool
+
+type ErrNotExist struct {
+	Ident identity.Stream
+}
+
+func (e ErrNotExist) Error() string {
+	return fmt.Sprintf("no item for stream %s", e.Ident)
+}
+
+func NotExist(err error) bool {
+	var notExist ErrNotExist
+	return errors.As(err, &notExist)
+}
+
+func Missing(id identity.Stream) error {
+	return ErrNotExist{Ident: id}
+}
+
+func Exists(id identity.Stream, ok bool) error {
+	if !ok {
+		return Missing(id)
+	}
+	return nil
+}
+
+// Map defines a collection of items tracked by a stream-id and the operations
+// on it
+type Map[T any] interface {
+	Load(context.Context, identity.Stream) (T, error)
+	Store(context.Context, identity.Stream, T) error
+	Delete(context.Context, identity.Stream)
```

Comment:
Similarly, for context, all of these operations *should* be fairly quick. I'm not sure if we'd need a context for say, context cancellation checking. I could potentially see it for context propagation of something like span contexts. But again, this is a fairly tiny operation. So wanting spans at that level or events at that level seems a bit extreme.